File: plugins/adaptive-learner-plugin-learning-repo/adaptive_learner_learning_repo/renderer.py
"""Orchestrator for the learning-repo renderer (Phase 42 / BL-30).

Public entry points:

  * :func:`load_context` — sync DB read, builds a
    :class:`RenderContext` for one project. Raises
    ``NotFoundError`` when the project doesn't exist.
  * :func:`render_repository` — high-level: load context, dispatch
    to every meta-file generator, build the topic folder stubs,
    return ``{path: content}`` for the whole tree.

No AI calls anywhere on this path. Renderer ignores
authorisation — the route layer (commit 4) is the right place
for per-user gating.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .context import RenderContext, derive_topics
from .labels import labels_for
from .meta import render_cheatsheet, render_readme, render_roadmap, render_stats
from .topic_folders import render_topic_folders

if TYPE_CHECKING:
    from sqlalchemy.orm import Session


def load_context(db: Session, project_id: str) -> RenderContext:
    """Build a :class:`RenderContext` for one project (sync read).

    ``app.*`` imports are function-local so the module stays
    importable in plugin-smoke-test contexts where the backend
    package isn't on sys.path. Matches the lazy-import pattern
    used by anki/notebooklm.
    """

    from sqlalchemy import select

    from app.exceptions import NotFoundError
    from app.models import (
        LearningProject,
        LearningSession,
        MethodSwitch,
        SessionNote,
        SessionRating,
        StepEvaluation,
    )

    project = db.get(LearningProject, project_id)
    if project is None:
        raise NotFoundError(f"LearningProject {project_id} not found")
    sessions = tuple(
        db.scalars(select(LearningSession).where(LearningSession.project_id == project_id))
    )
    session_ids = [s.id for s in sessions]
    ratings = tuple(
        db.scalars(select(SessionRating).where(SessionRating.session_id.in_(session_ids)))
        if session_ids
        else ()
    )
    step_evaluations = tuple(
        db.scalars(select(StepEvaluation).where(StepEvaluation.session_id.in_(session_ids)))
        if session_ids
        else ()
    )
    notes = tuple(
        db.scalars(select(SessionNote).where(SessionNote.session_id.in_(session_ids)))
        if session_ids
        else ()
    )
    method_switches = tuple(
        db.scalars(select(MethodSwitch).where(MethodSwitch.project_id == project_id))
    )
    topics = derive_topics(sessions)
    return RenderContext(
        project=project,
        sessions=sessions,
        ratings=ratings,
        step_evaluations=step_evaluations,
        method_switches=method_switches,
        notes=notes,
        topics=topics,
    )


def render_repository(
    db: Session,
    project_id: str,
    language: str = "en",
) -> dict[str, str]:
    """Build the full ``{path: content}`` map for one project."""

    ctx = load_context(db, project_id)
    return render_from_context(ctx, language)


def render_from_context(ctx: RenderContext, language: str = "en") -> dict[str, str]:
    """Dispatch to every generator. Pure function of the context.

    Split out so tests can build a context fixture and exercise
    the full dispatch without a DB round-trip.
    """

    labels = labels_for(language)
    tree: dict[str, str] = {
        "README.md": render_readme(ctx, labels),
        "LEARNING_STATS.md": render_stats(ctx, labels),
        "CHEATSHEET.md": render_cheatsheet(ctx, labels),
        "ROADMAP.md": render_roadmap(ctx, labels),
    }
    tree.update(render_topic_folders(ctx, labels))
    return tree

File: frontend/src/lib/learning-repo/renderer.ts
/**
 * Orchestrator for the learning-repo renderer — TypeScript
 * port of ``plugins/.../learning-repo/adaptive_learner_learning_repo/
 * renderer.py`` (Phase 49D / v1.32.0 / PHASE-42-STORAGE-
 * ABSTRACTION-01).
 *
 * Pure function of the context. Dispatches to every meta-file
 * generator + the topic-folder stubs, returns a
 * ``{path: content}`` map with the same shape as the backend's
 * ``/api/plugins/learning-repo/render/{project_id}`` response.
 *
 * No I/O. The DexieStorage learningRepo namespace (49E) calls
 * ``loadDexieContext`` (49B) to build the context, then hands
 * it here. The parity test (49F) loads a JSON fixture, builds
 * a context with a pinned ``rendered_at``, and asserts each
 * key in the returned map equals the corresponding golden
 * Markdown file byte-for-byte.
 */

import {labelsFor} from "./labels";
import type {RenderContext} from "./render-context";
import {renderCheatsheet} from "./renderers/cheatsheet";
import {renderReadme} from "./renderers/readme";
import {renderRoadmap} from "./renderers/roadmap";
import {renderStats} from "./renderers/stats";
import {renderTopicFolders} from "./topic-folders";

/**
 * Build the full ``{path: content}`` map for one project's
 * rendered repository tree.
 *
 * - Top-level files: ``README.md``, ``LEARNING_STATS.md``,
 *   ``CHEATSHEET.md``, ``ROADMAP.md``.
 * - Topic folders: ``NN_slug/README.md`` per ``TopicSlice``
 *   (empty when the project has no ``cycle_topics`` history).
 *
 * Caller supplies the language; ``labelsFor`` loads the bundled
 * i18n catalog and overlays ``repo.*`` strings onto the
 * English defaults.
 */
export async function renderRepository(
    ctx: RenderContext,
    language: string = "en",
): Promise<Record<string, string>> {
    const labels = await labelsFor(language);
    const tree: Record<string, string> = {
        "README.md": renderReadme(ctx, labels),
        "LEARNING_STATS.md": renderStats(ctx, labels),
        "CHEATSHEET.md": renderCheatsheet(ctx, labels),
        "ROADMAP.md": renderRoadmap(ctx, labels),
    };
    const topicStubs = renderTopicFolders(ctx, labels);
    for (const [path, content] of Object.entries(topicStubs)) {
        tree[path] = content;
    }
    return tree;
}
